### PR TITLE
parse tls scts

### PR DESCRIPTION
**Pull request by zakird**

### CODE REVIEW

File: zlib/grabber.go [hunk 1]
```diff
@@ -363,8 +366,8 @@ func makeGrabber(config *Config) func(*Conn) error {
 		if config.TLSExtendedRandom {
 			c.SetExtendedRandom()
 		}
-		if config.GatherSessionTicket {
-			c.SetGatherSessionTicket()
+		if config.SignedCertificateTimestampExt {
+			c.SetSignedCertificateTimestampExt()
```
Comment (dadrian) [inline zlib/grabber.go]: Please put that back.

File: ztools/ztls/common.go [hunk 2]
```diff
@@ -83,6 +83,7 @@ const (
 	extensionNextProtoNeg         uint16 = 13172 // not IANA assigned
 	extensionRenegotiationInfo    uint16 = 0xff01
 	extensionExtendedRandom       uint16 = 0x0028 // not IANA assigned
+	extensionSCT                  uint16 = 18
```
Comment (dadrian) [inline ztools/ztls/common.go]: It's 18, not 0x18?

File: ztools/ztls/handshake_messages.go [hunk 3]
```diff
@@ -652,9 +681,24 @@ func (m *serverHelloMsg) marshal() []byte {
 		z[1] = byte(extensionExtendedMasterSecret & 0xff)
 		z = z[4:]
 	}
+	if sctLen > 0 {
+		z[0] = byte(extensionSCT >> 8)
+		z[1] = byte(extensionSCT)
+		l := sctLen + 2
```
Comment (dadrian) [inline ztools/ztls/handshake_messages.go]: Are you sure this isn't plus 4? The length of `sctLen` includes the length of the length bytes?

File: ztools/ztls/handshake_messages.go [hunk 4]
```diff
@@ -762,6 +807,30 @@ func (m *serverHelloMsg) unmarshal(data []byte) bool {
...
+			d = d[2:]
+			if len(d) != l || l == 0 {
+				return false
+			}
+			m.scts = make([][]byte, 0, 3)
+			for len(d) != 0 {
+				if len(d) < 2 {
+					return false
+				}
+				sctLen := int(d[0])<<8 | int(d[1])
```
Comment (dadrian) [inline ztools/ztls/handshake_messages.go]: You treat `sctlen` different in `serverHello.unmarshal` vs `serverHello.marshal`. Why is there only a loop in one of these functions?

File: ztools/ztls/ztls_handshake.go [hunk 5]
```diff
@@ -28,18 +30,24 @@ type ClientHello struct {
...
+	SessionID                   []byte            `json:"session_id"`
+	CipherSuite                 CipherSuite       `json:"cipher_suite"`
+	CompressionMethod           uint8             `json:"compression_method"`
+	OcspStapling                bool              `json:"ocsp_stapling"`
+	TicketSupported             bool              `json:"ticket"`
+	SecureRenegotiation         bool              `json:"secure_renegotiation"`
+	HeartbeatSupported          bool              `json:"heartbeat"`
+	ExtendedRandom              []byte            `json:"extended_random,omitempty"`
+	ExtendedMasterSecret        bool              `json:"extended_master_secret"`
+	SignedCertificateTimestamps []parsedAndRawSCT `json:"scts,omitempty"`
```
Comment (dadrian) [inline ztools/ztls/ztls_handshake.go]: `parsedAndRawSCT` should be public since this struct is public.

File: ztools/ztls/ztls_handshake.go [hunk 6]
```diff
@@ -28,18 +30,24 @@ type ClientHello struct {
 	SessionID      []byte `json:"session_id,omitempty"`
 }
 
+type parsedAndRawSCT struct {
```
Comment (dadrian) [inline ztools/ztls/ztls_handshake.go]: This should be public, and the entries should be `,omitempty`.

File: ztools/ztls/ztls_handshake.go [hunk 7]
```diff
@@ -231,6 +239,17 @@ func (m *serverHelloMsg) MakeLog() *ServerHello {
 		sh.ExtendedRandom = make([]byte, len(m.extendedRandom))
 		copy(sh.ExtendedRandom, m.extendedRandom)
 	}
+	if len(m.scts) > 0 {
+		for _, rawSCT := range m.scts {
+			var out parsedAndRawSCT
+			copy(out.Raw, rawSCT)
```
Comment (dadrian) [inline ztools/ztls/ztls_handshake.go]: You need to allocate space for `Raw` before you copy, otherwise 0 bytes will be copied.

Comment (dadrian) [inline ztools/ztls/ztls_handshake.go]: Or cheat and `out.Raw = append(out.Raw, rawSCT...)`

File: ztools/ztls/ztls_handshake.go (hunk 5 above)
Comment (zakird) [inline ztools/ztls/ztls_handshake.go]: done

File: ztools/ztls/ztls_handshake.go (hunk 6 above)
Comment (zakird) [inline ztools/ztls/ztls_handshake.go]: done

File: ztools/ztls/ztls_handshake.go (hunk 7 above)
Comment (zakird) [inline ztools/ztls/ztls_handshake.go]: done

File: ztools/ztls/common.go (hunk 2 above)
Comment (zakird) [inline ztools/ztls/common.go]: yes

File: ztools/ztls/handshake_messages.go (hunk 3 above)
Comment (zakird) [inline ztools/ztls/handshake_messages.go]: that comes directly from golang stable

File: ztools/ztls/handshake_messages.go (hunk 4 above)
Comment (zakird) [inline ztools/ztls/handshake_messages.go]: THere are loops in both

**Pull request merged**